% This is "sig-alternate.tex" V2.0 May 2012
% This file should be compiled with V2.5 of "sig-alternate.cls" May 2012
%
% This example file demonstrates the use of the 'sig-alternate.cls'
% V2.5 LaTeX2e document class file. It is for those submitting
% articles to ACM Conference Proceedings WHO DO NOT WISH TO
% STRICTLY ADHERE TO THE SIGS (PUBS-BOARD-ENDORSED) STYLE.
% The 'sig-alternate.cls' file will produce a similar-looking,
% albeit, 'tighter' paper resulting in, invariably, fewer pages.
%
% ----------------------------------------------------------------------------------------------------------------
% This .tex file (and associated .cls V2.5) produces:
%       1) The Permission Statement
%       2) The Conference (location) Info information
%       3) The Copyright Line with ACM data
%       4) NO page numbers
%
% as against the acm_proc_article-sp.cls file which
% DOES NOT produce 1) thru' 3) above.
%
% Using 'sig-alternate.cls' you have control, however, from within
% the source .tex file, over both the CopyrightYear
% (defaulted to 200X) and the ACM Copyright Data
% (defaulted to X-XXXXX-XX-X/XX/XX).
% e.g.
% \CopyrightYear{2007} will cause 2007 to appear in the copyright line.
% \crdata{0-12345-67-8/90/12} will cause 0-12345-67-8/90/12 to appear in the copyright line.
%
% ---------------------------------------------------------------------------------------------------------------
% This .tex source is an example which *does* use
% the .bib file (from which the .bbl file % is produced).
% REMEMBER HOWEVER: After having produced the .bbl file,
% and prior to final submission, you *NEED* to 'insert'
% your .bbl file into your source .tex file so as to provide
% ONE 'self-contained' source file.
%
% ================= IF YOU HAVE QUESTIONS =======================
% Questions regarding the SIGS styles, SIGS policies and
% procedures, Conferences etc. should be sent to
% Adrienne Griscti (griscti@acm.org)
%
% Technical questions _only_ to
% Gerald Murray (murray@hq.acm.org)
% ===============================================================
%
% For tracking purposes - this is V2.0 - May 2012

\documentclass{sig-alternate}
\usepackage{epstopdf}
\usepackage{algorithm2e}
%\usepackage{auto-pst-pdf}
\begin{document}
%
% --- Author Metadata here ---
\conferenceinfo{GECCO2014}{2014 Victoria, BC, Canada}
%\CopyrightYear{2007} % Allows default copyright year (20XX) to be over-ridden - IF NEED BE.
%\crdata{0-12345-67-8/90/01}  % Allows default copyright data (0-89791-88-6/97/05) to be over-ridden - IF NEED BE.
% --- End of Author Metadata ---

\title{Enhancing the Performance of Genetic Algorithms in Combinatorial Optimization of Large and Difficult Problems}
\subtitle{Using Variable Adaptive Mutation Rates Controlled by 'Inbreeding'}
%
% You need the command \numberofauthors to handle the 'placement
% and alignment' of the authors beneath the title.
%
% For aesthetic reasons, we recommend 'three authors at a time'
% i.e. three 'name/affiliation blocks' be placed beneath the title.
%
% NOTE: You are NOT restricted in how many 'rows' of
% "name/affiliations" may appear. We just ask that you restrict
% the number of 'columns' to three.
%
% Because of the available 'opening page real-estate'
% we ask you to refrain from putting more than six authors
% (two rows with three columns) beneath the article title.
% More than six makes the first-page appear very cluttered indeed.
%
% Use the \alignauthor commands to handle the names
% and affiliations for an 'aesthetic maximum' of six authors.
% Add names, affiliations, addresses for
% the seventh etc. author(s) as the argument for the
% \additionalauthors command.
% These 'additional authors' will be output/set for you
% without further effort on your part as the last section in
% the body of your article BEFORE References or any Appendices.

\numberofauthors{4} %  in this sample file, there are a *total*
% of EIGHT authors. SIX appear on the 'first-page' (for formatting
% reasons) and the remaining two appear in the \additionalauthors section.
%
\author{
% You can go ahead and credit any number of authors here,
% e.g. one 'row of three' or two rows (consisting of one row of three
% and a second row of one, two or three).
%
% The command \alignauthor (no curly braces needed) should
% precede each author name, affiliation/snail-mail address and
% e-mail address. Additionally, tag each line of
% affiliation/address with \affaddr, and tag the
% e-mail address with \email.
%
% 1st. author
\alignauthor
Daniel Smullen\titlenote{Corresponding author.}\\
       \affaddr{UOIT}\\
       \affaddr{2000 Simcoe St. N}\\
       \affaddr{Oshawa, ON, Canada}\\
       \email{daniel.smullen@uoit.net}
% 2nd. author
\alignauthor
Jonathan Gillett\\
       \affaddr{UOIT}\\
       \affaddr{2000 Simcoe St. N}\\
       \affaddr{Oshawa, ON, Canada}\\
       \email{jonathan.gillett@uoit.net}
% 3rd. author
\alignauthor
Joseph Heron\\
       \affaddr{UOIT}\\
       \affaddr{2000 Simcoe St. N}\\
       \affaddr{Oshawa, ON, Canada}\\
       \email{joseph.heron@uoit.net}
\and  % use '\and' if you need 'another row' of author names
% 4th. author
\alignauthor Shahryar Rahnamayan\\
       \affaddr{UOIT}\\
       \affaddr{2000 Simcoe St. N}\\
       \affaddr{Oshawa, ON, Canada}\\
       \email{shahryar.rahnamayan@uoit.ca}
}
% There's nothing stopping you putting the seventh, eighth, etc.
% author on the opening page (as the 'third row') but we ask,
% for aesthetic reasons that you place these 'additional authors'
% in the \additional authors block, viz.
\additionalauthors{}
% Just remember to make sure that the TOTAL number of authors
% is the number that will appear on the first page PLUS the
% number that will appear in the \additionalauthors section.

\maketitle
\begin{abstract}
Blah blah blah, here is our abstract.
\end{abstract}

\keywords{Genetic Algorithms, Combinatorial Optimization, N Queens Problem, Variable Mutation}




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 
% INTRODUCTION
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
Genetic algorithms serve an important role in various applications, but particularly useful areas are those where they can perform stochastic generation of solutions for combinatorial problems. One notable example of this is the N-Queens problem. One approach to solving the N-Queens problem using a genetic algorithm is to implement chromosomes which model each position of each queen on the chessboard per axis. The permuted values stores in the chromosome represent the sequential row positions of each of the queens, encoded into binary values. Thus, the column positions increase by one for each queen and the intersections (referred to as collisions) are calculated per row or on the diagonal to determine fitness. As with any genetic algorithm, evaluating the fitness of the chromosomes in a generation is required to determine if a solution has been generated. By tuning the parameters of the genetic algorithm, performance can increase and decrease based on the problem encountered. We have found that by replacing a static tuned mutation value with an adaptive value controlled by chromosome similarity, better performance can be achieved than using static values, particularly when there is no \emph{a priori} knowledge about the ideal static mutation rate. This is compounded when the solution landscape is large, such as in higher-order N-Queens problems.

\subsection{Background and Problem Domain}
The N-Queens problem becomes orders of magnitude complex as the number of queens and the size of the chessboard increases. Finding all solutions is a simple but nontrivial problem. When two or more queens share a row or diagonal, a collision occurs for each pair of queens, meaning that the board state is not a solution. The Eight-Queens problem is the classical version of the puzzle, and even with this configuration the problem is computationally expensive. There are $64 \choose 8$ arrangements of the queens on a standard $8\times{}8$ board, while deterministic evaluation has shown that there are only 92 distinct solutions. Even including optimizations such as constraining a queen to a single row, there are still $8^8$ possible distinct arrangements. Naturally, in higher-order versions of the problem, combinatorial explosion occurs. The Nine-Queens problem has 352 distinct solutions. The Ten-Queens problem has 724. Table \ref{table:numuniquesol} on page \pageref{table:numuniquesol} shows the growth of the number of solutions. While the number of solutions is known for problems involving up to 26 queens, there is currently no known formula for the exact number of distinct solutions. That is to say, higher order N-Queens problems are intractable. Deterministic methods for approaching these problems are largely useless, and the problem remains unsolved.

- (ADD CITATIONS!)
  
% CITATIONS EXAMPLE WITH ALL CITATATIONS
- Cite me!\cite{crawford1992solving,homaifar1992e1,andrews2006investigation,tuson1998adapting, wolpert1997no,srinivas1994adaptive,goldberg1988genetic}

\subsubsection{The N-Queens Puzzle}

- Based on the complexity of the problem explain how the overhead of using a bruteforce 
  approach is infeasible for larger N Queens and the justification for stochastic
  methods.(ADD CITATIONS!)

- Emphasize that our application of the problem is finding distinct solutions,
  since this is much harder than just finding solutions (many of which may be
  duplicates in other random stochastic methods).


\subsubsection{Motivation}
- The motive for variable mutation
    --> We were not able initially (without using reflection \& rotation like we
        are now) to get all 92 solutions to 8 queens (impossible with fixed mutation to solve
        in a reasonable amount of time)
    
    --> Started looking at biology, were inspired by the fact that nature has
        a natural ordering to prevent inbreeding within organisms (e.g. purebreed
        dogs often have more health problems, easy citation!)
        
    --> Attempted to come up with a way to apply the negative affects of inbreeding 
        that occur in nature to solve our GA problem
        
    --> Our var. mutation solution worked exceptionally well, solved all 92
        solutions nearly instantaneously, motivated us to pursue the implications
        further.


\subsection{Related Work}

- This is where you can cite all of the references Joseph found regarding other types of variable mutation.
- Joseph can help you out with this section




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 
% APPROACH
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Our Approach}

- Describe the variable mutation rate algorithm and the two parameters that it
  has in comparison to fixed mutation's one param, (step size for changing the
  mutation rate and the inbreeding threshold used to adjust the mutation rate).


\subsection{Applying Variable Mutation Rates}

- For explaining the variable mutation rate there is no need for a formal algorithm
  essentially it boils down to the following:

  - For each generation evaluate the population chromose similarity (see below
    for the population chromosome similiarity algorithm)

  - If the population chromosome similarity is less than the threshold increase
    the mutation rate by the step size. If the similarity is greater than the
    threshold decrease the mutation rate by the step size. This results in
    the mutation rate adjusting up or down such that the population chromosome
    similarity approaches an equilibrium near the inbreeding threshold.


% This could be combined with the previous subsection as this was a section you had added before
% but can be explained by demonstation the similiarity algorithm and explaining the 
% inbreeding threshold
%\subsubsection{How Adaptation Occurs}


% This is essentially explaining the chromosome similarity algroithm since the concept of 
% applying variable mutation from the previous subsection can be easily explained in words 
% without a formal algorithm.
\subsubsection{Chromosome Similarity Algorithm}

- For explaining the chromosome similarity algorithm (which is a little more
  complicated to avoid the naiive O(n!) solution, instead it is now linear
  complexity). Even though in terms of theoretical big-O complexity it is
  constant (since it's just based on the population size which is constant) it
  is worthwhile mentioning that from our empirical evidence using profiling of
  the code using the naiive O(n!) solution has a significant impact for larger
  population sizes (so much so that it caused a significant degrade in performance
  motivating us to profile the code and improve the original algorithm and create
  the following linear complexity algorithm)\cite{bowman:reasoning}


\begin{algorithm}
  \SetKwProg{Fn}{Function}{}{end}\SetKwFunction{Similarity}{Similarity}%
  \SetKwData{Similar}{similar}\SetKwData{Value}{value}\SetKwData{Matched}{matched}\SetKwData{Length}{length}\SetKwArray{Sorted}{sorted}
  \SetKwFunction{Sort}{Sort}
  \SetKwInOut{Input}{input}\SetKwInOut{Output}{output}
  \SetAlgoLined
  \DontPrintSemicolon
  
  \Fn{\Similarity{$chromosomes$}}{
  
  \Input{An array of $chromosomes$}
  \Output{Fraction of chromosomes that are similar}
  \BlankLine
  
  \Similar $\leftarrow$ 0\;
  \Matched $\leftarrow$ false\;
  \Length $\leftarrow$ length of $chromosomes$\;
  \BlankLine
  
  \tcp{Sort using an arbitrary sorting algorithm}
  \Sorted $\leftarrow$ \Sort{$chromosomes$}\;
  \BlankLine
  
  \For{$i \leftarrow 0$ \KwTo $\Length -1$}
  {
    \uIf{\Sorted{$i$} == \Sorted{$i + 1$}}
    {
      \Similar $\leftarrow$ \Similar + 1\;
      \Matched $\leftarrow$ true\;
    }
    \ElseIf{\Matched}
    {
      \Similar $\leftarrow$ \Similar + 1\;
      \Matched $\leftarrow$ false\;
    }
    \BlankLine
    
    \tcp{Case where the last item is a match}
    \If{\Matched and $\left(i + 1 == \Length - 1 \right)$}
    {
      \Similar $\leftarrow$ \Similar + 1\;
    }
  }
  \BlankLine
  \KwRet{\Similar / \Length}
  }
\caption{Chromosome similarity function}
\end{algorithm}




% I thought it would be important to add a section on implementation
% this could be condensed, but I think it is very important to
% explain how we actually implemented and applied the GA process
\subsection{Implementation}

- An overview of our GA implementation so that someone wishing to replicate our 
  results knows how we implemented our solution. 

- Could potentially provide a URL to the source code implementation on github, I think
  this would be a very good idea.


% This is the design of the chromosome, which is important to the implementation details
\subsubsection{Chromosome Design}
- Describe the chromosome design, the ordering of each gene of the chromosome represents
  the horizontal position of each queen on the board and the corresponding vertical
  position of the queen is an integer stored in the gene (array index) of the chromosome.

  --> I think it would be wortwhile to have a very small diagram highlighting this that shows
      the correlation between a chromosome to the queens on a small (4x4 or 5x5) chessboard

% The implementation detail of the fitness funciton including the algorithm
\subsubsection{Fitness Function}
- Fitness function, each pair of queens that have a vertical or diagonal collision is 
  recorded as a value of 2 (one collision from the perspective of each queen). This number 
  is then represented  as a fraction over 1 such that a chromosome with a larger number of
  collisions will have a much lower fitness value.
  

\begin{algorithm}
  \SetKwProg{Fn}{Function}{}{end}\SetKwFunction{Fitness}{Fitness}%
  \SetKwData{Collisions}{collisions}\SetKwData{Length}{length}\SetKwArray{Chromosome}{$chromosome$}\SetKwData{Yi}{$y_i$}\SetKwData{Yj}{$y_j$}
  \SetKwFunction{Abs}{abs}
  \SetKwInOut{Input}{input}\SetKwInOut{Output}{output}
  \SetAlgoLined
  \DontPrintSemicolon
  
  \Fn{\Fitness{$chromosome$}}
  {
    \Input{A single $chromosome$}
    \Output{A fitness value for the chromosome}
    \BlankLine
    
    \Collisions $\leftarrow$ 0\;
    \Length $\leftarrow$ length of the $chromosome$\;
    \BlankLine
    
    \For{$i \leftarrow 0$ \KwTo $\Length -1$}
    {
      \tcp{Check each gene against the current}
      $j \leftarrow \left( i + 1 \right) \mod{\Length}$\;
      \While{$j$ != $i$}
      {
        \Yi $\leftarrow \Chromosome{i}$\;
        \Yj $\leftarrow \Chromosome{j}$\;
        \BlankLine
        
        \tcp{Check for vertical collision}
        \If{\Yi == \Yj}
        {
          \Collisions $\leftarrow \Collisions + 1$\; 
        }
        \BlankLine
        
        \tcp{Check for diagonal collision}
        \If{\Abs{$\left(i - j \right)$ / $\left(\Yi - \Yj \right)$} == 1}
        {
          \Collisions $\leftarrow \Collisions + 1$\;
        }
        \BlankLine
        
        $j \leftarrow j + 1$\;
        $j \leftarrow j \mod{\Length}$\;
      }
    }
    \BlankLine

    \eIf{\Collisions == 0}
    {
      \KwRet{1}\;
    }
    {
      \KwRet{1 / \Collisions}\;
    }
  }
\caption{Fitness function}
\end{algorithm}


- If no collisions occur the default value is 1 resulting in a fitness of (1/1 = 1) 
  which is a solution to the problem. 
  
- For queens with multiple collisions each pair of queens will be recorded as 2 collisions, 
  in the event that there are three pairs of queens with collisions, one pair with a vertical 
  collision, one with a diagonal, and one with a separate diagonal collision the fitness value 
  would be (1 / 6).
  
- In the event where there are multiple types of collisions (say two pairs of queens each with 
  vertical collisions, and one with a diagonal collision) the fitness value would be (1 / 8), 
  for a higher number of collisions this often makes the fitness value even lower than 
  just a multiple of 2 x pairs of collisions, the fitness value decreases more than linearly
  for a higher number of collisions.

  --> Might be helpful to give a diagram

  --> Due to the ordering of the queens in the chromosome there can only be vertical or diagonal
      collisions (since two queens can never be on the same row).


% The selection method, we use roulette wheel, however the implementation details of how
% roulette wheel is implemented is important
\subsubsection{Selection Method}
- Roulette wheel selection method, by generating a random floating point number and selecting
  a chromosome value where the random number lies within the bounds. The range
  of random values that can be chosen changes based on the maximum upper and lower
  bounds of the sum of the fitness of the chromosomes in the population.

- Each of the ranges of values that a particular chromosome can have is weighted
  based on the fitness of the chromosome. For example in a population with 4
  chromosomes two of which are solutions (fitness 1) and two with a single pair of collisions
  (fitness 1/2 = 0.5) the ranges of values for each chromosome would be as
  follows:

  [0, 1), [1, 2), [2, 2.5), [2.5, 3), and the bounds of the random floating point value
  generated for the roulette wheel selection would be [0, 3).


% The operations, these are pretty standard but we need to cite why we chose 70% and 30%
% we used these because they were recommended (e.g. from AI lectures) but need textbooks
% or papers to suggest why we used them
\subsubsection{Chromosome Operations}
- Crossover operation (70\% chance), cloning (30\% chance), (cite paper justifying why
  we used these values, they are often recommended/used values)

- Background mutation operation, applied to chromosomes, this is variable rather
  than a traditional fixed value, e.g. a 5\% value results in a 5\% chance of
  mutation being applied to chromosomes.

- Mutation operator changes ONE of the genes of the chromosome randomly, which
  results in changing the the y coordinate of a random queen in the chromosome 
  to a random value within the range of possible y values.

- Mutation operator DOES NOT result in an invalid chromosome, it is limited to the
  range of possible (valid) y values.


% How we evaluate chromosomes in the population and determine distinct solutions
% and apply rotate and reflection to find additional solutions
% The use of terminology here is important DISTINCT implies that the particular
% arrangment of queens is different from another rotation/reflection of the same original
% solution. UNIQUE would consider all reflections/rotations as ONE "unique" solution
\subsubsection{Chromosome Evaluation}
- If a chromsome of the current population has a fitness value of 1 it is compared 
  to the list of previous solutions to see if it is unique. If the solution is 
  distinct the rotation (rotating the solution an additional three times) and 
  reflection (performing reflection on the solution) followed by three additional
  rotations operations are applied to find a total of 8 solutions. Each solution 
  found after rotation and reflection is compared to the list of previous solutions to 
  verify that it is a distinct solution.

  --> We do not keep duplicate solutions, this is important since someone may think
      that out of the 1000's of solutions we found many of them are just duplicates,
      when in fact they are all DISTINCT solutions.

- The current population is then replaced with the new population that was created
  by applying the crossover, cloning, and mutation operations.


% I would instead call this methodology and make testing/validation/etc. subsections
%\subsubsection{Testing and Validation}
\subsection{Methodology}

- Implementation in Java

- Experiments were conducted on the HPC facilities provided by SHARCNET

- describe the study and control (diff. fixed mutation rates, list them all, vs.
  variable mutation rate with a fixed inbreeding threshold of 15\%), this had
  1 fixed param (inbreeding threshold)
  
  --> This should probably be put in a table. I'll create one for you


- N queens problem sizes used: For 8 - 16 queens used each n queens size, for
  16 - 26 used each even sized N queens problem, lastly a test using 32 queens.

- Number of generations for each N queens problem (10 million generations for
  all except 32 queens), 50 million generations were used for 32 queens given
  the increased complexity of the problem.

- sample sizes (30, 15, 10)

    --> I will give you a table with each n queens problem, the sample sizes used 
        for each and the number of generations, this would reduce a lot of text
        needed to explain the items.

    --> justify why the sample sizes were reduced for larger problems given the 
        computational limitations of SHARCNET (max 256 jobs, 7 days CPU time, 
        regardless and your job would be killed). We would like larger sample sizes
        (100+ runs) but given the CPU time limitations of SHARCNET could not.

- describe the data collected, explain how each of the attributes such as
  population fitness, similarity are calculated based on the mean of the population
  for 1000 generations at a time (mean of means), rather than the mean of each
  population for each generation (TOO MUCH DATA!)



% This really should be part of results, but could be left here
% to me it really doesn't fit in with approach since explaining
% our approach can't really justify yet "Why not fixed mutation"
\subsection{Why Not Fixed Mutation?}

- This should be part of the end of results basically summarizing the pros
  and cons of fixed mutation and variable with reference to how our results
  show that in our case (for certain larger problems) variable mutation was better.




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 
% RESULTS
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% I would call this results, but performance analysis is good too
\section{Results}

- cite specific results and why we think they are important, what is the significance
  of them and how they support our results/hypothesis that in the case of N Queens
  var. mutation is better than fixed.

- cite the result showing the best fixed mutation vs. the variable mutation
  for each N-queens problem, try and use both the figure and the table, the table
  has some additional interesting information which cannot be conveyed in the
  image alone.
  
  --> I will create the table for you, essentially showing each N queens problem, the
      best results for fixed mutation vs. the best results for variable mutation

- cite interesting results of the fat boxplots in the range of chromosome similarity
  for the optimal fixed mutation rates, use the "person stepping" analogy and how
  that certain fixed mutation rates had the widest range in chromosome similiarty
  allowing it to hone in on solutions faster.

- Try and incorporate one of the scatter plots as well that show very interesting
  results and see if you can use it to compare/contrast the results of the 
  following plots
  
    - variable mutation rate scatter plot
    - variable mutation rate similarity scatter plot
    - best fixed mutation rate similarity scatter plot




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 
% CONCLUSIONS
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Conclusions}

- Wait until the rest of the paper \& we have more feedback before
  writing the conclusions.

- Further research into adjusting the inbreeding threshold, for the purpose
  of the research a constant inbreeding threshold of 15\% was used, however
  further research could be done in testing different thresholds.

- Comparing the results of variable mutation with fixed mutation using other
  types of combinatorial/optimization problems such as TSP, constraint
  satisfaction problem (CSP), etc.

- Variable population size based on the amount of inbreeding (if you have
  a lot of inbreeding in nature the organisms will have higher mutatation
  rate and deformities, the population will shrink)

- Having the mutation operator affect more than one gene if it goes > 100\%? 

\begin{flushleft}\end{flushleft}


%\end{document}  % This is where a 'short' article might terminate

%ACKNOWLEDGMENTS are optional
\section{Acknowledgments}
This section is optional; it is a location for you
to acknowledge grants, funding, editing assistance and
what have you.

- Shahryar

- SHARCNET


%
% The following two commands are all you need in the
% initial runs of your .tex file to
% produce the bibliography for the citations in your paper.
\bibliographystyle{abbrv}
\bibliography{sigproc}  % sigproc.bib is the name of the Bibliography in this case
% You must have a proper ".bib" file
%  and remember to run:
% latex bibtex latex latex
% to resolve all references
%
% ACM needs 'a single self-contained file'!
%

%APPENDICES are optional
%\balancecolumns
\appendix
%Appendix A
\section{Results}
The rules about hierarchical headings discussed above for
the body of the article are different in the appendices.
In the \textbf{appendix} environment, the command
\textbf{section} is used to
indicate the start of each Appendix, with alphabetic order
designation (i.e. the first is A, the second B, etc.) and
a title (if you include one).  So, if you need
hierarchical structure
\textit{within} an Appendix, start with \textbf{subsection} as the
highest level. Here is an outline of the body of this
document in Appendix-appropriate form:

\subsection{Tables}
\begin{table}
\centering
\caption{Number of unique solutions to the N-Queens Problem}
\begin{tabular}{|l|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|} \hline
N & 8 & 9 & 10 & 11 & 12 & 13 & 14 & 15 & 16 & 17 & 18 & 19 & 20 & 21 & 22 & 23 & 24 & 25 & 26\\ \hline
Distinct Solutions & 92 & 352 & 724 & 2680 & 14200 & 73712 & 365596& 2279184& 14772512& 95815104& 666090624& 4968057848& 39029188884& 314666222712& 2691008701644& 24233937684440& 227514171973736& 2207893435808352& 22317699616364044\\
\hline\end{tabular}
\label{table:numuniquesol}
\end{table}
- Additional tables can go here

\subsection{Figures}

- Additional figures can go here

\subsection{Source Code}

- Could link to github and possibly the github wiki, or explanation
  of the R analysis source code.

\subsection{Raw Data}

- Link to the dropbox URL containing the actual data from SHARCNET used for the research


\end{document}
